rust/src/read.rs

```diff
@@ -698,6 +684,63 @@ impl<'a> Iterator for MessageStream<'a> {
     }
 }
 
+/// Like [`RawMessageStream`], but constructs a [`Message`]
+/// (complete with its [`Channel`]) from the raw header and data.
+///
+/// This stops at the end of the data section and does not read the summary.
+///
+/// Because tying the lifetime of each message to the underlying MCAP memory map
+/// makes it very difficult to send between threads or use in async land,
+/// and because we assume _most_ MCAP files have _most_ messages in compressed chunks,
+/// yielded [`Message`]s have unbounded lifetimes.
+/// For messages we've decompressed into their own buffers, this is free!
+/// For uncompressed messages, we take a copy of the message's data.
+pub struct MessageStream<'a> {
+    inner: RawMessageStream<'a>,
+}
+
+impl<'a> MessageStream<'a> {
+    pub fn new(buf: &'a [u8]) -> McapResult<Self> {
+        Self::new_with_options(buf, enum_set!())
+    }
+
+    pub fn new_with_options(buf: &'a [u8], options: EnumSet<Options>) -> McapResult<Self> {
+        RawMessageStream::new_with_options(buf, options).map(|inner| Self { inner })
+    }
+}
+
+impl<'a> Iterator for MessageStream<'a> {
+    type Item = McapResult<Message<'static>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.inner.next() {
+            Some(Ok((header, data))) => {
+                // Messages must have a previously-read channel.
+                let channel = match self.inner.channeler.get(header.channel_id) {
+                    Some(c) => c,
+                    None => {
+                        return Some(Err(McapError::UnknownChannel(
+                            header.sequence,
+                            header.channel_id,
+                        )))
+                    }
+                };
+
+                Some(Ok(Message {
+                    channel,
+                    sequence: header.sequence,
+                    log_time: header.log_time,
+                    publish_time: header.publish_time,
+                    data: Cow::Owned(data.into_owned()),
+                }))
+            }
+            // Coerce Option<McapResult<(header, data)>> into Option<McapResult<Message>>
```

Comment:
i feel like this comment should cover the match arm above?

Reply:
I was trying to explain (maybe unsuccessfully :sweat_smile:) why we need these other two arms (which don't seem to do anything) instead of some catch-all
```rust
otherwise => otherwise
```
which complains that the result types don't match:
```
   --> src/read.rs:737:26
    |
716 |           match self.inner.next() {
    |           ----------------------- `match` arms have incompatible types
...
729 | /                 Some(Ok(Message {
730 | |                     channel,
731 | |                     sequence: header.sequence,
732 | |                     log_time: header.log_time,
733 | |                     publish_time: header.publish_time,
734 | |                     data: Cow::Owned(data.into_owned()),
735 | |                 }))
    | |___________________- this is found to be of type `std::option::Option<Result<Message<'static>, McapError>>`
736 |               },
737 |               otherwise => otherwise
    |                            ^^^^^^^^^ expected struct `Message`, found tuple
    |
    = note: expected enum `std::option::Option<Result<Message<'static>, _>>`
               found enum `std::option::Option<Result<(MessageHeader, Cow<'_, [u8]>), _>>````
```
Something like
```
            otherwise => otherwise.map(|r| r.map(|_never_ok| unreachable!())),
```
would work instead, but that seems a lot less clear to me.

Suggestions on a better comment to get that across here?

Reply:
maybe something like `// Cover remaining match arms with trivial implementations`?